File: formula-cs/Formula/ShuntingYard/Variable.cs
namespace Formula.ShuntingYard;

public delegate ResolvedValue VariableResolver(IDataContext context, string key);

public readonly record struct Variable(string Key, VariableResolver VariableResolver) : INode
{
    public ResolvedValue Get(IDataContext context)
    {
        return VariableResolver(context, Key);
    }
}

File: java/formula-java/src/main/java/org/formula/parse/shuntingyard/Variable.java
package org.formula.parse.shuntingyard;

import org.formula.Resolvable;
import org.formula.ResolvedValue;
import org.formula.context.DataContext;

public record Variable(String key, VariableResolver variableResolver) implements Node {

    public ResolvedValue get(DataContext context) {
        return variableResolver.resolve(context, key);
    }
}
